### PR TITLE
parse spec info for generated libraries

### PR DESCRIPTION
**Pull request by devoncarew**
- no longer write generated libraries that wouldn't contain any code
- make the process to clean up previously generated files more lightweight (instead of deleting all the source at the start of generation, we only remove files after generation that hadn't been modified)
- vend in the node `web-specs` package; parse that as part of generation in order to read spec info related to the generated libraries (we don't use this info as yet, but could use it to prune which specs we choose to generate for - see https://github.com/dart-lang/web/issues/105)

This PR is sort of a grab-bag; thoughts welcome.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Review by kevmoo — APPROVED** [review]
LOVE IT! should wait for 👍  from @srujzs , though

File: tool/bindings_generator/filesystem_api.dart [hunk 1]
```diff
@@ -33,5 +43,19 @@ extension FileSystemExtension on FileSystem {
     }
   }
 
+  @JS('readFileSync')
```
Comment (srujzs) [inline tool/bindings_generator/filesystem_api.dart:46]: I believe you don't need both, you can make the second argument optional.

Calls to external JS interop functions only pass in the arguments in the call-site, so you still need the two different calls to `readFileSync` if you don't want to pass `null` for `options`, but you don't need both external members.

**Review by srujzs — APPROVED** [review]
Nice, thanks! While ideally we should probably move the addition of the web spec to a PR that uses it, I'm totally okay with keeping the processing of the web spec here.

File: tool/bindings_generator/filesystem_api.dart (hunk 1 above)
Comment (devoncarew) [reply tool/bindings_generator/filesystem_api.dart:46]: I switch this to just:

```
external JSAny readFileSync(JSString path, [JSReadFileOptions options]);
```

and it seems to work fine. Thanks! That's much simpler.

**Comment by devoncarew** [comment]
> Nice, thanks! While ideally we should probably move the addition of the web spec to a PR that uses it,

Yeah, totally - this would normally be the 1st half of a PR. After looking at the info we were parsing however, it didn't immediately jump out at me which libraries not to include in the generation. I think this will need someone w/ more domain expertise to incrementally prune the generated libraries.

We may even want to use the info to generate some libraries into a `src/experimental` directory. It could be nice to make some APIs available, but w/o strong guarantees wrt api stability.

**Review by srujzs — APPROVED** [review]

**Comment by srujzs** [comment]
> We may even want to use the info to generate some libraries into a src/experimental directory. It could be nice to make some APIs available, but w/o strong guarantees wrt api stability.

This would be great! One of the concerns I had was that due to the noise of some of the in-progress API definitions, we'd be releasing breaking changes too often. Moving them to a folder where we make no guarantees of stability should hopefully minimize those breaking changes.

**Comment by kevmoo** [comment]
Turns out that info is NOT as useful as I'd hoped. It's too complex to explain here...

**Pull request merged**